torch/_inductor/config.py

```diff
@@ -728,6 +728,9 @@ class aot_inductor:
     # rather than embedded into the data section. Needed to support 1B+ parameter models
     force_mmap_weights: bool = False
 
+    # Flag to determine if an input tensor, which is also an output tensor and is being mutated, should be mutated in the graph.
+    keep_inference_input_mutations: bool = False
```

Comment:
Don't we want to do this always?  I'd expect this to be the current behavior before this PR.  It at least is for the Python wrapper code.

Can you provide an example?

Reply:
Yes, this is the current behavior. I introduced this configuration to facilitate the support of ATen operations that have the `out` parameter. Initially, my plan was to set this configuration to `True` specifically for any ATen operation that includes an `out` parameter. Currently, the `out` parameter remains unchanged which conflicts with the ATen operations.

Reply:
If your program mutates its inputs -- you can't remove that.  That would be incorrect.  If the cpp_wrapper is removing this, that is just a bug that should be fixed rather than a config flag.

Reply:
Can you give an example of a program where we are discarding input mutations?

Reply:
For now, I think the input mutations for AOT compilation always rely on `_unlift_graph`/`_unlift`. But it only supports the inputs and outputs are the alias of model buffers or parameters. @bdhirsh , please correct me.
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L162-L201

Take `aten::clamp.out(Tensor self, Scalar? min=None, Scalar? max=None, *, Tensor(a!) out) -> Tensor(a!)` as an example

```python
out_tensor = torch.randn((128, 128), dtype=torch.float, device=device)
inp_tensor = torch.randn((128, 128), dtype=torch.float, device=device)
torch.clamp(input=inp_tensor, min=0.05, max=0.05, out=out_tensor)
```

The graph produced by `aot_export_module` is as follows during AOT compilation. But, `_unlift_graph` cannot correct the produced graph to align with the semantics of the ATen operations that contain the `out` parameter.
```python
opcode         name       target                  args                       kwargs
-------------  ---------  ----------------------  -------------------------  --------
placeholder    arg0_1     arg0_1                  ()                         {}
placeholder    arg1_1     arg1_1                  ()                         {}
call_function  clamp_min  aten.clamp_min.default  (arg0_1, 0.05)             {}
call_function  clamp_max  aten.clamp_max.default  (clamp_min, 0.05)          {}
output         output     output                  ((clamp_max, clamp_max),)  {}
```

### Additional Context

The partial of invocation sequence of `torch._inductor.aot_compile` is `compile_fx_aot` -> `compile_fx` -> `aot_export_module` -> `_aot_export_function`. 

The `_aot_export_function` builds a `AOTConfig` instance w/ `keep_inference_input_mutations=False` and then passes the `AOTConfig` instance to `create_aot_dispatcher_function`.
https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1350
```python
    # The export use case doesn't care about several bits of AOTConfig
    # (1) compilers (we just export the graph)
    # (2) partitioners (export is only full graph, user can partition themselves)
    aot_config = AOTConfig(
        fw_compiler=None,
        bw_compiler=None,
        inference_compiler=None,
        partition_fn=None,
        decompositions=decompositions,
        num_params_buffers=num_params_buffers,
        aot_id=next(AOT_COUNTER),
        # For now there's no use case involving keeping input mutations in the graph
        # (which we can only do in the inference case anyway).
        # We can add this later if we need to.
        keep_inference_input_mutations=False,
        dynamic_shapes=dynamic_shapes,
        aot_autograd_arg_pos_to_source=None,
        is_export=True,
        no_tangents=no_tangents,
        pre_dispatch=pre_dispatch,
    )
```

Since the `keep_inference_input_mutations` is `False`, the mutation type of inputs will be always `MutationType.MUTATED_OUT_GRAPH` when collecting mutation metadata.
https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/schemas.py#L147-L167
```python
    @functools.cached_property
    def mutation_type(self) -> MutationType:
        if (
            (not self.mutates_data)
            and (not self.mutates_metadata)
            and not (self.mutation_inductor_storage_resize)
        ):
            return MutationType.NOT_MUTATED

        if _check_if_mutation_can_be_in_graph(
            self.keep_input_mutations,
            self.mutates_data,
            self.mutates_metadata,
            self.mutations_hidden_from_autograd,
            self.mutations_under_no_grad_or_inference_mode,
            self.mutates_storage_metadata,
            self.mutation_inductor_storage_resize,
            self.requires_grad,
        ):
            return MutationType.MUTATED_IN_GRAPH

        return MutationType.MUTATED_OUT_GRAPH
```

Eventually, the graph will be produced as follows according to the mutation metadata analysis.
```python
opcode         name       target                  args                       kwargs
-------------  ---------  ----------------------  -------------------------  --------
placeholder    arg0_1     arg0_1                  ()                         {}
placeholder    arg1_1     arg1_1                  ()                         {}
call_function  clamp_min  aten.clamp_min.default  (arg0_1, 0.05)             {}
call_function  clamp_max  aten.clamp_max.default  (clamp_min, 0.05)          {}
output         output     output                  ((clamp_max, clamp_max),)  {}
```

When the graph gets produced, `_unlift_graph` will be invoked to `unlift` the inputs and outputs to refine the graph according to the graph signature.
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L1451-L1458
```python
    if V.aot_compilation is True:
        with functorch_config.patch(unlift_effect_tokens=True):
            gm, graph_signature = aot_export_module(
                model_,
                example_inputs_,
                trace_joint=False,
                decompositions=decompositions,
            )
        unlifted_gm = _unlift_graph(model_, gm, graph_signature)
```

Unfortunately, `_unlift_graph` only handles the buffers and parameters of model: `graph_signature.inputs_to_parameters`, `graph_signature.inputs_to_buffers` and `graph_signature.buffers_to_mutate`. 
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L162-L190
```python
    # In AOTI, module parameters and buffers are not lifted as graph inputs.
    # As a result, mutation to buffers has side effect which makes their initial
    # values different from Eager. So we clone them here as a copy.
    # We are not cloning for parameters, although it will be needed if we want to
    # support training.
    for node in placeholder_nodes:
        node_name = node.name
        if node_name in graph_signature.inputs_to_parameters:
            parameter_name = graph_signature.inputs_to_parameters[node_name]
            lifted_inputs.append(parameter_name)
        elif node_name in graph_signature.inputs_to_buffers:
            buffer_name = graph_signature.inputs_to_buffers[node_name]
            lifted_inputs.append(buffer_name)
            gm.meta[
                get_cloned_parameter_buffer_name(buffer_name)
            ] = clone_preserve_strides(state_dict[buffer_name])
        else:
            assert node_name in graph_signature.user_inputs
            lifted_inputs.append(None)

    from torch.export._unlift import _unlift

    outputs = list(gm.graph.nodes)[-1].args[0]
    mutated_outputs = []
    for out in outputs:
        if out.name in graph_signature.buffers_to_mutate:
            mutated_outputs.append(graph_signature.buffers_to_mutate[out.name])
        else:
            mutated_outputs.append(None)
```

However, the single aten operation to be compiled does not have such attributes and not mention to the `out` is neither model `buffers` not model `parameters`.

Reply:
@jansel , I'm not sure if I have given the example of what you want to get. Please let me know if I need to give more information.

Reply:
IMO they should both keep mutations.  Discarding mutation is a correctness issue and should never happen.

Reply:
@bdhirsh , may I know your input? I just updated this PR and added the logic to handle this case in `_unlift`. But it looks like not straightforward to raise the semantics to support `out`. In addition, I still keep `keep_inference_input_mutations` implementation for your comparison. 

https://github.com/pytorch/pytorch/pull/124926/files#diff-5f8f15b69738f6bd565a99e6b29b8b379ff11f899a154c20c148c6ecbc183b41R68-R93